diff --git a/src-android/Reflex/Dom/Android/MainWidget.hsc b/src-android/Reflex/Dom/Android/MainWidget.hsc
index d2349e1..fada5b8 100644
--- a/src-android/Reflex/Dom/Android/MainWidget.hsc
+++ b/src-android/Reflex/Dom/Android/MainWidget.hsc
@@ -57,7 +57,7 @@ startMainWidget a url jsm = do
         \     return JSON.parse(jsaddle.syncMessage(JSON.stringify(a)));\n\
         \ });\n\
         \ return function(batch) {\n\
-        \  core.processReq(batch);\n\
+        \  core.processReqs(batch);\n\
         \ };\n\
         \})();\n\
         \jsaddle.postReady();\n"
